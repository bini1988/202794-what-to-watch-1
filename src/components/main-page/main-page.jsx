import React, {PureComponent} from "react";
import PropTypes from "prop-types";
import {connect} from "react-redux";

import {Operation} from "../../reducer/movies/movies";
import {getPromoMovie} from "../../reducer/movies/selectors";

import {Operation as GenresOperation, ActionCreator as GenresActionCreator} from "../../reducer/genres/genres";
import {getActiveGenre, getMoviesGenresList, getMoviesByGenre, getLimitByGenre} from "../../reducer/genres/selectors";

import MovieCard from "../movie-card/movie-card";
import PageHeader from "../page-header/page-header";
import PageTitle from "../page-title/page-title";
import PageFooter from "../page-footer/page-footer";
import UserBlock from "../user-block/user-block";
import MoviesCatalog from "../movies-catalog/movies-catalog";

class MainPage extends PureComponent {
  render() {
    const {
      promoMovieCard = {},
      movies,
      moviesGenres,
      activeGenre,
      moviesLimit,
      onGenreChange,
      onMoviesMore,
      onToMyListToggle
    } = this.props;

    return (
      <React.Fragment>
        <MovieCard
          card={promoMovieCard}
          onToMyListToggle={onToMyListToggle}>
          <MovieCard.Header component={PageHeader}>
            <PageTitle hidden>{`WTW`}</PageTitle>
            <UserBlock/>
          </MovieCard.Header>
          <MovieCard.InfoWrapper>
            <MovieCard.Poster/>
            <MovieCard.Description>
              <MovieCard.PlayButton/>
              <MovieCard.ListButton/>
            </MovieCard.Description>
          </MovieCard.InfoWrapper>
        </MovieCard>
        <div className="page-content">
          <MoviesCatalog
            movies={movies}
            moviesGenres={moviesGenres}
            activeGenre={activeGenre}
            onGenreChange={onGenreChange}
            onMoviesMore={onMoviesMore}
            limit={moviesLimit}/>
          <PageFooter/>
        </div>
      </React.Fragment>
    );
  }

  componentDidMount() {
    this.props.onMoviesFetch();
    this.props.onPromoMovieFetch();
  }
}

MainPage.propTypes = {
  /** Текущий промо фильм */
  promoMovieCard: PropTypes.object,
  /** Список отображаемых фильмов */
  movies: MoviesCatalog.propTypes.movies,
  /** Список отображаемых жанров фильмов */
  moviesGenres: MoviesCatalog.propTypes.moviesGenres,
  /** Активный жанр фильмов */
  activeGenre: MoviesCatalog.propTypes.activeGenre,
  /** Изменить фильтр списка фильмов по жанру */
  onGenreChange: MoviesCatalog.propTypes.onGenreChange,
  /** Количество отображаемых фильмов */
  moviesLimit: PropTypes.number,
  /** Получить список фильмов */
  onMoviesFetch: PropTypes.func,
  /** Получить следующий набор фильмов соответвующего жанра */
  onMoviesMore: PropTypes.func,
  /** Получить текущий промо фильм */
  onPromoMovieFetch: PropTypes.func,
  /** Добавить/удалить фильм из списока «к просмотру» */
  onToMyListToggle: PropTypes.func,
  /** Вложенные элементы */
  children: PropTypes.any,
};

const mapStateToProps = (state) => {
  const activeGenre = getActiveGenre(state);

  return {
    promoMovieCard: getPromoMovie(state),
    movies: getMoviesByGenre(state, activeGenre),
    moviesGenres: getMoviesGenresList(state),
    moviesLimit: getLimitByGenre(state, activeGenre),
    activeGenre,
  };
};
const mapDispatchToProps = {
  onGenreChange: GenresActionCreator.changeActiveGenre,
  onMoviesMore: GenresOperation.increaseGenreLimit,
  onMoviesFetch: Operation.fetchMovies,
  onPromoMovieFetch: Operation.fetchPromoMovie,
  onToMyListToggle: Operation.toggleMovieToMyList,
};

export {MainPage};
export default connect(mapStateToProps, mapDispatchToProps)(MainPage);

