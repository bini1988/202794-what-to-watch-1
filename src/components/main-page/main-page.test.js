import React from "react";
import {MemoryRouter} from "react-router-dom";
import {Provider} from "react-redux";
import configureStore from "redux-mock-store";
import renderer from "react-test-renderer";
import {MainPage} from "./main-page";
import genreGroupsMock from "../../mocks/movies-groups";
import promoMovie from "../../mocks/movie-card";

const initialState = {};
const mockStore = configureStore();

it(`MainPage correctly renders default markup`, () => {
  const tree = renderer.create(
      <Provider store={mockStore(initialState)}>
        <MemoryRouter
          initialEntries = {[`/`]}>
          <MainPage
            promoMovieCard={promoMovie}
            movies={genreGroupsMock[`All Genres`]}
            moviesGenres={Object.keys(genreGroupsMock)}
            onGenreChange={() => {}}
            onMoviesFetch={() => {}}
            onPromoMovieFetch={() => {}}
            onToMyListAdd={() => {}}
          />
        </MemoryRouter>
      </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
