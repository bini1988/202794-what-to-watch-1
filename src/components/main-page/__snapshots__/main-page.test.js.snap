// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPage correctly renders default markup 1`] = `
Array [
  <section
    className="movie-card"
    style={
      Object {
        "background": "#9B7E61",
      }
    }
  >
    <header
      className="page-header movie-card__head"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="movie-card__bg"
      >
        <img
          alt="War of the Worlds"
          src="https://es31-server.appspot.com/wtw/static/film/background/War_of_the_Worlds.jpg"
        />
      </div>
      <h1
        className="page-title visually-hidden"
      >
        WTW
      </h1>
      <div
        className="user-block"
      >
        <a
          className="user-block__link"
          href="/login"
          onClick={[Function]}
        >
          Sign in
        </a>
      </div>
    </header>
    <div
      className="movie-card__wrap"
    >
      <div
        className="movie-card__info"
      >
        <div
          className="movie-card__poster"
        >
          <img
            alt="War of the Worlds"
            height="327"
            src="https://es31-server.appspot.com/wtw/static/film/poster/War_of_the_Worlds.jpg"
            width="218"
          />
        </div>
        <div
          className="movie-card__desc"
        >
          <h2
            className="movie-card__title"
          >
            War of the Worlds
          </h2>
          <p
            className="movie-card__meta"
          >
            <span
              className="movie-card__genre"
            >
              Adventure
            </span>
            <span
              className="movie-card__year"
            >
              2005
            </span>
          </p>
          <div
            className="movie-card__buttons"
          >
            <a
              className="btn btn--play movie-card__button"
              href="/film/1/player"
              onClick={[Function]}
            >
              <svg
                height={19}
                viewBox="0 0 19 19"
                width={19}
              >
                <use
                  xlinkHref="#play-s"
                />
              </svg>
              <span>
                Play
              </span>
            </a>
            <button
              className="btn btn--list movie-card__button"
              type="button"
            >
              <svg
                height={20}
                viewBox="0 0 19 20"
                width={19}
              >
                <use
                  xlinkHref="#add"
                />
              </svg>
              <span>
                My list
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
  </section>,
  <div
    className="page-content"
  >
    <section
      className="catalog"
    >
      <h2
        className="catalog__title visually-hidden"
      >
        Catalog
      </h2>
      <ul
        className="catalog__genres-list"
      >
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            onClick={[Function]}
          >
            All Genres
          </a>
        </li>
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            onClick={[Function]}
          >
            Genre #1
          </a>
        </li>
        <li
          className="catalog__genres-item "
        >
          <a
            className="catalog__genres-link"
            onClick={[Function]}
          >
            Genre #2
          </a>
        </li>
      </ul>
      <div
        className="catalog__movies-list"
      >
        <article
          className="small-movie-card catalog__movies-card"
          id="movie-11"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              height={175}
              muted={true}
              onCanPlayThrough={[Function]}
              onEnded={[Function]}
              onLoadStart={[Function]}
              onPause={[Function]}
              onTimeUpdate={[Function]}
              poster="https://es31-server.appspot.com/wtw/static/film/preview/war-of-the-worlds.jpg"
              preload="none"
              src="http://media.xiph.org/mango/tears_of_steel_1080p.webm"
              width={280}
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/film/11"
              onClick={[Function]}
            >
              War of the Worlds
            </a>
          </h3>
        </article>
        <article
          className="small-movie-card catalog__movies-card"
          id="movie-22"
          onMouseEnter={[Function]}
          onMouseLeave={[Function]}
        >
          <div
            className="small-movie-card__image"
          >
            <video
              height={175}
              muted={true}
              onCanPlayThrough={[Function]}
              onEnded={[Function]}
              onLoadStart={[Function]}
              onPause={[Function]}
              onTimeUpdate={[Function]}
              poster="https://es31-server.appspot.com/wtw/static/film/preview/war-of-the-worlds.jpg"
              preload="none"
              src="http://media.xiph.org/mango/tears_of_steel_1080p.webm"
              width={280}
            />
          </div>
          <h3
            className="small-movie-card__title"
          >
            <a
              className="small-movie-card__link"
              href="/film/22"
              onClick={[Function]}
            >
              War of the Worlds
            </a>
          </h3>
        </article>
      </div>
    </section>
    <footer
      className="page-footer"
    >
      <div
        className="logo"
      >
        <a
          className="logo__link logo__link--light"
          href="/"
          onClick={[Function]}
        >
          <span
            className="logo__letter logo__letter--1"
          >
            W
          </span>
          <span
            className="logo__letter logo__letter--2"
          >
            T
          </span>
          <span
            className="logo__letter logo__letter--3"
          >
            W
          </span>
        </a>
      </div>
      <div
        className="copyright"
      >
        <p>
          Â© 2019 What to watch Ltd.
        </p>
      </div>
    </footer>
  </div>,
]
`;
